h/auth/policy.py

```diff
@@ -41,34 +43,104 @@ def forget(self, request):
+        if proxy_userid is not None:
+            return proxy_userid
+
+        # username from BasicAuth header
+        return self.basic_auth_policy.unauthenticated_userid(request)
 
     def authenticated_userid(self, request):
         """
-        Currently the behavior here is different than other Auth policies
-        as the net result is that, although Authentication may well be
-        successful, we don't end up with an authenticated ``userid``.
+        Return any forwarded userid or None
 
-        Thus, ``request.authenticated_userid`` will be None and
-        ```request.user`` will not be set (see :py:mod:`h.accounts`)
+        Rely mostly on ``Pyramid.authentication.BasicAuthAuthenticationPolicy``'s
+        authenticated_userid, but don't
+        actually return a userid unless there is a forwarded user (the auth
+        client itself is not a "user")
 
-        However, the request will be authenticated. In a view,
-        ``effective_principals=security.Authenticated`` would be satisfied.
+        Although this looks as if it trusts the unauthenticated_userid irrationally,
+        rest assured that the authenticating check callback will always be called
+        (via ``Pyramid.authentication.BasicAuthAuthenticationPolicy.callback``)
+        before any non-None value is returned.
 
-        .. todo::
+        :rtype: `~h.models.user.User.userid` or None
+        """
+        forwarded_userid = AuthClientPolicy.forwarded_userid(request)
 
-           Extend me here (carefully) to operate on behalf of user in authority)
+        # only evaluate setting an authenticated_userid if forwarded user is present
+        if forwarded_userid is None:
+            return None
+
+        # username extracted from BasicAuth header
+        auth_userid = self.basic_auth_policy.unauthenticated_userid(request)
+
+        # authentication of BasicAuth and forwarded user—this will invoke check
+        callback_ok = self.basic_auth_policy.callback(auth_userid, request)
+
+        if callback_ok is not None:
+            return forwarded_userid  # This should always be a userid, not an auth_client id
+
+    def effective_principals(self, request):
         """
-        return None
+        Return a list of principals for the request
+
+        This will concatenate the principals returned by the registered
+        ``check`` callback (which is a list or None) with Pyramid's
+        system principal(s).
+
+        If the ``check`` callback returns None—that is, if authentication is
+        unsuccessful—the returned principals will only contain Pyramid's
+        ``system.Everyone`` principal (and Pyramid will not consider the
+        request as authenticated).
+
+        :rtype: list ``['system.Everyone']`` concatenated with any principals
+                from a successful authentication, if any
+        """
+        return self.basic_auth_policy.effective_principals(request)
```

Comment:
I took a look at re-implementing `effective_principals` instead of proxying to `BasicAuthAuthenticationPolicy`'s variant (inherited through `CallbackAuthenticationPolicy`), but there's just enough Pyramid magic going on there that it seems unwise.